### PR TITLE
Fix `--inputs` 1 — `call` command

### PR DESCRIPTION
**Pull request by kasperski95**
Use low-level Starknet.py utils to handle data transformer calls properly. Create `CallUseCase` to avoid adding even more code to the `GatewayFacade`. Show raw and transformed results.

Related https://github.com/software-mansion/protostar/issues/1220

### CODE REVIEW

**Comment by MaksymilianDemitraszek** [comment]
Can you link the PR to the related issue? 🥺 🙏

**Comment by kasperski95** [comment]
> Can you link the PR to the related issue? 🥺 🙏

This PR is already linked. (Check the bottom of the PR's description.)

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
Can we split the fix and refactors 🥺 🙏? those should not be interleaving, I am okay with discussing refactors later as a separate task.

File: protostar/starknet_gateway/call/call_protocols.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from abc import abstractmethod
+from typing import Optional, Protocol
+
+from protostar.starknet import Address, AbiType
+
+from .call_structs import CallPayload, CallResponse
+
+
+class ClientProtocol(Protocol):
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/call/call_protocols.py:9]: Avoiding circular dependencies. I can probably remove it in favor of importing types conditionally with TYPE_CHECKING.

File: protostar/starknet_gateway/call/call_protocols.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    @abstractmethod
+    async def send_call(self, payload: CallPayload) -> CallResponse:
+        ...
+
+
+class AbiResolverProtocol(Protocol):
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/call/call_protocols.py:15]: Resolving in favor of the thread above.

File: protostar/commands/call/call_command_messages.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        lines.append(fmt.bold(str(self.call_output.cairo_data)))
+        lines.append("")
+        lines.append("[" + fmt.colorize("CYAN", "TRANSFORMED RESULT") + "]")
+        lines.append(fmt.bold(self._get_response_as_json()))
+        return "\n".join(lines)
+
+    def format_dict(self) -> dict:
+        return {
+            "raw": self.call_output.cairo_data,
+            "transformed": self.call_output.human_data,
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/call/call_command_messages.py]: [add output suffix](https://github.com/software-mansion/protostar/pull/1275/commits/b9548644bf219d6dcd2fc58e920195d6a973e4c5)

File: protostar/starknet_gateway/call/call_protocols.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/call/call_protocols.py]: [remove protocols](https://github.com/software-mansion/protostar/pull/1275/commits/458cc42ed77d3c1ca6ffac81ba39987ee43eee30)

File: tests/integration/gateway/uint256_as_input_test.py [hunk 4]
```diff
@@ -6,7 +6,6 @@
 from tests.data.contracts import CONTRACT_WITH_UINT256_CONSTRUCTOR
 from tests.integration.conftest import CreateProtostarProjectFixture
 from tests.integration.protostar_fixture import ProtostarFixture
-
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply tests/integration/gateway/uint256_as_input_test.py:9]: [bring back empty line](https://github.com/software-mansion/protostar/pull/1275/commits/6218e605e1f2810acf17b34582368a4fd69121f8)

File: protostar/starknet_gateway/call/call_exceptions.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+from protostar.protostar_exception import ProtostarException
+
+
+class AbiNotFoundException(ProtostarException):
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/call/call_exceptions.py]: [remove exception](https://github.com/software-mansion/protostar/pull/1275/commits/5cb0d7d240a5c7fbbb5757a5a9efe2d68680d382)

File: protostar/starknet_gateway/abi_resolver.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+)
+
+from protostar.starknet import AbiType, Address
+
+
+class AbiResolver:
+    def __init__(self, client: Client) -> None:
+        self._client = client
+
+    async def resolve(self, address: Address) -> Optional[AbiType]:
```
Comment (MaksymilianDemitraszek) [inline protostar/starknet_gateway/abi_resolver.py:22]: I would remove this class and replaced it with `get_abi` function, it much better represents the purpose. Resolve is internal starknet py name and in our context it is not obvious what this class does.

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
It is much better than before, but it still can be simplified further

File: protostar/commands/call/call_command.py [hunk 7]
```diff
@@ -79,34 +64,33 @@ def arguments(self):
...
+        abi_resolver = AbiResolver(client=gateway_client)
+        call_use_case = CallUseCase(client=gateway_facade, abi_resolver=abi_resolver)
+        response = await call_use_case.execute(
+            CallInput(
+                address=contract_address,
+                selector=Selector(function_name),
+                inputs=inputs,
+                abi=None,
+            )
         )
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/call/call_command.py:95]: Business logic shouldn't be here. `Command` is just a presentation layer.

File: protostar/starknet_gateway/call/call_use_case.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        self,
+        abi_resolver: "AbiResolver",
+        client: "GatewayFacade",
+    ) -> None:
+        self._client = client
+        self._abi_resolver = abi_resolver
+
+    async def execute(self, input_data: CallInput) -> CallOutput:
+        cairo_calldata = await self._transform_calldata_if_necessary(input_data)
+        payload = CallPayload(
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/call/call_use_case.py:30]: Ok. I can replace structs with function arguments.

File: protostar/starknet_gateway/abi_resolver.py (hunk 6 above)
Comment (kasperski95) [reply protostar/starknet_gateway/abi_resolver.py:22]: Resolver is a generic name, not only specific to Starknet.py. This class finds a proper contract and then fetches the ABI. You can suggest a better for the class.

File: protostar/starknet_gateway/call/call_structs.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+@dataclass
+class CallPayload:
+    address: Address
+    selector: Selector
+    cairo_calldata: CairoDataRepresentation
+
+
+@dataclass
+class CallResponse:
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/call/call_structs.py]: [remove structs in favor of args](https://github.com/software-mansion/protostar/pull/1275/commits/3555e348c8e8e6cf5a0001001ffa9ed2ddd609fd)

File: protostar/starknet_gateway/call/call_use_case.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/call/call_use_case.py]: [remove structs in favor of args](https://github.com/software-mansion/protostar/pull/1275/commits/3555e348c8e8e6cf5a0001001ffa9ed2ddd609fd)

File: protostar/starknet_gateway/call/call_structs.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
+from dataclasses import dataclass
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/call/call_structs.py:1]: I'd rather not. There are also types there. Merging will make the use case file messy.

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/starknet_gateway/call/call_structs.py:1]: Imo those are thightly coupled to UseCase and shouldn't be placed in a separate module. It is easier to follow this way.

File: protostar/starknet_gateway/abi_resolver.py (hunk 6 above)
Comment (MaksymilianDemitraszek) [reply protostar/starknet_gateway/abi_resolver.py:22]: It shouldn't be a class it should be a function

File: protostar/starknet_gateway/call/call_use_case.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            address=input_data.address,
+            selector=input_data.selector,
+            cairo_calldata=cairo_calldata,
+        )
+        response = await self._client.send_call(payload)
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/starknet_gateway/call/call_use_case.py]: Discussed on slack, leave it as it is

File: protostar/starknet_gateway/abi_resolver.py (hunk 6 above)
Comment (MaksymilianDemitraszek) [reply protostar/starknet_gateway/abi_resolver.py:22]: Discussed on slack, leave it as it is

**Review by MaksymilianDemitraszek — APPROVED** [review]

**Pull request merged**